File: 725.split-linked-list-in-parts.cpp
/*
 * @lc app=leetcode id=725 lang=cpp
 *
 * [725] Split Linked List in Parts
 */
#include <vector>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    vector<ListNode *> splitListToParts(ListNode *head, int k)
    {
        vector<ListNode *> res = {head};
        int count = 0;
        ListNode *cur = head;
        while (cur)
        {
            ++count;
            cur = cur->next;
        }
        int group_size = count / k;
        int rest = count % k;
        cur = head;
        count = 0;
        while (cur)
        {
            ++count;
            if (group_size == 0)
            {
                auto next_cur = cur->next;
                cur->next = nullptr;
                cur = next_cur;
                res.push_back(cur);
            }
            else if (count % group_size == 0)
            {
                if (rest)
                {
                    cur = cur->next;
                    --rest;
                }
                auto next_cur = cur->next;
                cur->next = nullptr;
                cur = next_cur;
                res.push_back(cur);
            }
            else
            {
                cur = cur->next;
            }
        }
        res.resize(k);
        return res;
    }
};
// @lc code=end

File: 725.split-linked-list-in-parts.java
import java.util.List;

/*
 * @lc app=leetcode id=725 lang=java
 *
 * [725] Split Linked List in Parts
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    // public class ListNode {
    // int val;
    // ListNode next;

    // ListNode() {
    // }

    // ListNode(int val) {
    // this.val = val;
    // }

    // ListNode(int val, ListNode next) {
    // this.val = val;
    // this.next = next;
    // }
    // }

    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode[] res = new ListNode[k];
        int count = 0;
        ListNode cur = head;
        while (cur != null) {
            ++count;
            cur = cur.next;
        }
        int group_size = count / k;
        int rest = count % k;
        cur = head;
        int pos = 0;
        while (cur != null) {
            res[pos++] = cur;
            for (int i = 0; i < group_size - 1; ++i) {
                cur = cur.next;
            }
            if (group_size > 0 && rest > 0) {
                cur = cur.next;
                --rest;
            }
            ListNode next_cur = cur.next;
            cur.next = null;
            cur = next_cur;
        }
        return res;
    }
}
// @lc code=end
